### PR TITLE
Fixed #29538 -- Fixed crash of ordering by related fields when Meta.ordering contains expressions.

### PR DESCRIPTION
**Pull request by jerivas**
Also fixes https://code.djangoproject.com/ticket/33678
Supersedes #10146

### CODE REVIEW

File: django/db/models/expressions.py [hunk 1]
```diff
@@ -390,6 +390,18 @@ def relabeled_clone(self, change_map):
     def copy(self):
         return copy.copy(self)
 
+    def prefix_references(self, prefix):
+        clone = self.copy()
+        clone.set_source_expressions(
+            [
+                F(f"{prefix}__{expr.name}")
```
Comment (charettes) [inline django/db/models/expressions.py:397]: Minor but I think the `LOOKUP_SEP` should not be baked in `prefix_references` to allow caller to include it themselves.

```suggestion
                F(f"{prefix}{expr.name}")
```

File: django/db/models/sql/compiler.py [hunk 2]
```diff
@@ -907,6 +907,8 @@ def find_ordering_name(
 
             results = []
             for item in opts.ordering:
+                if hasattr(item, "prefix_references"):
```
Comment (charettes) [inline django/db/models/sql/compiler.py:910]: I'd suggest moving the `item.prefix_references` part to the `if isinstance(item, Orderby)` instead as all expressions are already wrapped in an `OrderBy` and the latter always has a `prefix_references` method so this conditional is unnecessary

```python
if isinstance(item, OrderBy):
    results.append((item.prefix_references(f"{name}{LOOKUP_SEP}"), False))
    continue
```

File: tests/ordering/models.py [hunk 3]
```diff
@@ -62,3 +62,17 @@ class Reference(models.Model):
...
+    name = models.CharField(max_length=30)
+
+    class Meta:
+        ordering = (models.functions.Lower("name"),)
+
+
+class OrderedByExpressionChild(models.Model):
+    parent = models.ForeignKey(OrderedByExpression, models.CASCADE)
+
+    def __str__(self) -> str:
```
Comment (charettes) [inline tests/ordering/models.py:77]: Unnecessary ->

```suggestion
    def __str__(self):
```

**Review by charettes — COMMENTED** [review]
Thanks for the patch @jerivas, left a few comments.

I think it would also be great to test ordering chaining with a leaf expression to make sure 


```python
class OrderedByExpressionGrandChild(models.Model):
    parent = models.ForeignKey(OrderedByExpressionChild, models.CASCADE)

OrderedByExpressionGrandChild.objects.order_by('parent')
```

I have a feeling the `name` usage is not appropriately in this case; we'd want to `order_by(Lower('parent__parent__name'))` in this case and not `Lower('parent__name')`.

**Comment by jerivas** [comment]
Thanks @charettes. I just pushed b00fbf832780bbef043fdfe43cb9a52cb4ac96ef to address the review.

Wouldn't the ordering of `OrderedByExpressionGrandChild.objects.order_by('parent')` fall back to insertion order since `OrderedByExpressionChild` has no ordering defined in `Meta`?

**Comment by charettes** [comment]
@jerivas you're right, could you also set `OrderedByExpressionChild.Meta.ordering = ('parent',)` to make sure the chain is complete?

**Comment by jerivas** [comment]
@charettes you're right, it breaks if we go down another level. Any pointers on how to fix it?

**Comment by charettes** [comment]
I suggest you adjust the recursive call into `find_ordering_name` to treat `isinstance(result, OrderBy)` differently by calling `prefix_references` on them.

**Comment by jerivas** [comment]
@charettes let me know what you think of https://github.com/django/django/pull/15666/commits/533c403661e9c573676f783ae3dc06b958e51dc8

**Comment by felixxm** [comment]
@jerivas Thanks :+1: Welcome aboard :boat:

**Pull request merged**